### ISSUE TITLE
Is there an easy way to concatenate tables that share the same primary key ?

### ISSUE THREAD

**Issue by healiseu** [yandex/ClickHouse#4137]
Hi, I have come across the following problem:

Let's take the following views that all share a common column hb1

```sql
SELECT hb1, supCountry FROM VW1 order by hb1
┌─hb1────┬─supCountry───┐
│  6     │ USA          │
│ 13     │ SPAIN        │
└────────┴──────────────┘


SELECT hb1, catTotal FROM VW3 order by hb1
┌─hb1────┬─catTotal─┐
│  6     │ 400      │
│  9     │ 200      │
│ 15     │ 200      │
└────────┴──────────┘


SELECT hb1, supName FROM VW2 order by hb1
┌─hb1────┬─supName───────────────┐
│  6     │ Acme Widget Suppliers │
│  9     │ Big Red Tool and Die  │
│ 13     │ Perfunctory Parts     │
│ 15     │ Alien Aircaft Inc.    │
└────────┴───────────────────────┘
```
I want to construct a table 
```sql
DESCRIBE TABLE Test

┌─name───────┬─type─────────────┬─default_type─┬─default_expression─┐
│ hb1        │ UInt32           │              │                    │
│ supCountry │ Nullable(String) │              │                    │
│ catTotal   │ Nullable(UInt16) │              │                    │
│ supName    │ Nullable(String) │              │                    │
└────────────┴──────────────────┴──────────────┴────────────────────┘
```
that will concatenate those columns above. The result set will be:

```sql
SELECT hb1, supCountry, catTotal, supName
FROM Test
ORDER BY hb1

┌─hb1─┬─supCountry─┬─catTotal─┬─supName───────────────┐
│   6 │ USA        │      400 │ Acme Widget Suppliers │
│   9 │ ᴺᵁᴸᴸ       │      200 │ Big Red Tool and Die  │
│  13 │ SPAIN      │     ᴺᵁᴸᴸ │ Perfunctory Parts     │
│  15 │ ᴺᵁᴸᴸ       │      200 │ Alien Aircaft Inc.    │
└─────┴────────────┴──────────┴───────────────────────┘
```
#### Question: Is there an efficient-simple way to concatenate those columns and to avoid writing a nested loop of joins ?

For example I was trying a solution with `ADD COLUMN` and `ReplacingMergeTree` without any luck. I was also thinking that since `ClickHouse` is a columnar database, there must be code that transforms columns to tuples but is that available as a function at SQL level ?

PS: Issue #528 might be relevant here

**Comment by alexey-milovidov** [comment]
You can write it with GROUP BY:

```
SELECT hb1, any(supCountry), any(catTotal), any(supName)
FROM
(
    SELECT hb1, supCountry, NULL AS catTotal, NULL AS supName FROM VW1
    UNION ALL
    SELECT hb1, NULL AS supCountry, catTotal, NULL AS supName FROM VW3
    UNION ALL
    SELECT hb1, NULL AS supCountry, NULL AS catTotal, supName FROM VW3
)
GROUP BY hb1
```

Though this is not the most efficient way.

**Comment by healiseu** [comment]
Thank @alexey-milovidov, I tested your code on a small table of 16 rows by 15 columns and with missing data. It takes TRIACLICK about 40msec to do the transformation. This is better than my `ANY LEFT OUTER JOIN` approach that took about 56msec to return the result. Anyway I expect this benchmark will rise to the order of seconds for millions of rows...

**Comment by alexey-milovidov** [comment]
You're right, the timings measured on just a few rows are misleading. You should measure on millions.
For less amount of rows, the query speed is dominated by various preparations than the data processing itself.

> I expect that the table can be reconstructed likewise the MergeTree() engine by reading and concatenating the filtered columns in much less time.

Yes, it could be theoretically implemented by merging of sorted data streams.

**Comment by healiseu** [comment]
Yes that's what I have read in the [`Block Streams`][1] section of the documentation. Anyway, the `TRIACLICK` release is going to be for experimental and demonstration purposes. I am aiming to show that Associative-Semiotic-Hypergraph approach can be advantageous for several reasons. There can be a lot of  things to discuss in terms of the implementation but my main goal is not performance but ultimate flexibility in dynamic data modeling and adhoc filtering. Nevertheless performance is dominant in my case too.

I have started my long journey from the data modeling perspective and that's where I will try to close the circle. In practice that means all the specifications for a new data modeling approach (R3DM/S3DM), a new kind of query language (CQL-Chain Query Language) and TRIACLICK the first implemented system that follows these.

[1]: https://clickhouse.yandex/docs/en/development/architecture/#block-streams

**Comment by healiseu** [comment]
The table (2,780,185 x 41) was loaded on TriaClick (i.e. ClickHouse Associative Semiotic Hypergraph engine) in approximately **63sec**, including optimizations, almost the same time like qlikview. The size of **TriaClick table engines on the disk was 507MB** (but data encoding was not optimal).

Then I applied a filter with low selectivity and the filtered set became **2,678,522 rows**. Finally I reconstructed a table with 5 columns (projection) using your `UNION ALL`- `GROUP BY` code and I fetched 100 rows only. The relational set was **constructed in 5sec** which is not bad for non-optimal performance. TriaClick system, according to `clickhouse-client`, **processed 16.42 million rows** to return this result set.
